### PR TITLE
Add example: "Filter csv records matching a predicate".

### LINKED ISSUE #230: add "filter csv records matching a predicate"
please see [csv tutorial](https://docs.rs/csv/*/csv/tutorial/index.html#pipelining)  if using this resource please remember to attribute original author

cc https://github.com/brson/rust-cookbook/issues/212

See [CONTRIBUTING.md](https://github.com/brson/rust-cookbook/blob/master/CONTRIBUTING.md) for details.

### PR DESCRIPTION
**Pull request by sotocodes**
Resolves #230

### CODE REVIEW

File: src/encoding.md [hunk 1]
```diff
@@ -543,6 +544,65 @@ Ashley-Zurich-92";
...
+        if record.iter().any(|field| field == &query) {
+            wtr.write_record(&record)?;
+        }
+    }
+
+    wtr.flush()?;
+    Ok(())
+}
+#
+# quick_main!(run);
```
Comment (somebrel) [inline src/encoding.md]: Travis is complaining about this line. Any ideas?

**Comment by somebrel** [comment]
Travis is complaining about this line `# quick_main!(run);`. Any ideas?

**Comment by budziq** [comment]
Hi. Travis is not complaining about the specific line.
"encoding_sect_filter_csv_records_matching_a_predicate_line_603" is the autogenerated test name (new feature of skeptic test runner).  And travis is just reporting that this test fails. Are the tests passing locally on your setup with `cargo test`? Likely one of the `?` int he code returned error.

**Comment by budziq** [comment]
@sotocodes the `bail!` results in returning from `run` early with an `Err` result. `quick_main` will unwrap the result and there is your panic. While the setup for cli app is sane it is non ideal for a testcase. 
We can either make the example `no_run` or change the example with some hardcoded query.

**Comment by sotocodes** [comment]
@budziq `cargo test` outputs
```shell
error: failed to run custom build command for `openssl-sys v0.9.15`
process didn't exit successfully: `/home/username/_Code/OSS/rust-cookbook/target/debug/build/openssl-sys-169fec58669c2940/build-script-build` (exit code: 101)
--- stdout
cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
cargo:rerun-if-env-changed=OPENSSL_DIR
run pkg_config fail: "`\"pkg-config\" \"--libs\" \"--cflags\" \"openssl\"` did not exit successfully: exit code: 1\n--- stderr\nPackage openssl was not found in the pkg-config search path.\nPerhaps you should add the directory containing `openssl.pc\'\nto the PKG_CONFIG_PATH environment variable\nNo package \'openssl\' found\n"

--- stderr
thread 'main' panicked at '

Could not find directory of OpenSSL installation, and this `-sys` crate cannot
proceed without this knowledge. If OpenSSL is installed and this crate had
trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
compilation process.

If you're in a situation where you think the directory *should* be found
automatically, please open a bug at https://github.com/sfackler/rust-openssl
and include information about your system as well as this message.

    $HOST = x86_64-unknown-linux-gnu
    $TARGET = x86_64-unknown-linux-gnu
    openssl-sys = 0.9.15

', /home/username/.cargo/registry/src/github.com-1ecc6299db9ec823/openssl-sys-0.9.15/build.rs:198
note: Run with `RUST_BACKTRACE=1` for a backtrace.

Build failed, waiting for other jobs to finish...
error: build failed
```
What do I need OpenSSL for?

**Comment by budziq** [comment]
openssl-sys (ffi wrapper over system OpenSSL) is a build predicate for hyper and reqwest that are used in some other examples. Basically you need to install OpenSSL developement package as described in 
 https://github.com/sfackler/rust-openssl/blob/master/README.md#building

**Comment by sotocodes** [comment]
Alright now I can run the tests, thanks @budziq. I changed the example to use a hardcoded query. It works well but the test still fails:

`test encoding_sect_filter_csv_records_matching_a_predicate_line_598 ... FAILED`

```rust
fn run() -> Result<()> {
    let query = "CA";
    let data = "\
City,State,Population,Latitude,Longitude
Kenai,AK,7610,60.5544444,-151.2583333
Oakman,AL,,33.7133333,-87.3886111
Sandfort,AL,,32.3380556,-85.2233333
West Hollywood,CA,37031,34.0900000,-118.3608333";

    let mut rdr = csv::ReaderBuilder::new().from_reader(data.as_bytes());
    let mut wtr = csv::Writer::from_writer(io::stdout());
    wtr.write_record(rdr.headers()?)?;

    for result in rdr.records() {
        let record = result?;
        if record.iter().any(|field| field == jjjjkkkjjjquery) {
            wtr.write_record(&record)?;
        }
    }

    wtr.flush()?;
    Ok(())
}
```

**Comment by budziq** [comment]
Hi I see two problems:

1. you have not pushed the new version so the travis test result is not updated
2. the code you have pasted has an error (I assume that it was introduced while pasting) 
```rust
error[E0425]: cannot find value `jjjjkkkjjjquery` in this scope
  --> src/main.rs:30:47
   |
30 |         if record.iter().any(|field| field == jjjjkkkjjjquery) {
   |                                               ^^^^^^^^^^^^^^^ not found in this scope
```

Please update the PR with newest code and if the problems persist I'll look at the problem :)

File: src/encoding.md [hunk 2]
```diff
@@ -543,6 +544,60 @@ Ashley-Zurich-92";
...
 
+[ex-csv-filter]: #ex-csv-filter
+<a name="ex-csv-filter"></a>
+## Filter CSV records matching a predicate
+
+[![csv-badge]][csv] [![cat-encoding-badge]][cat-encoding]
+
+_Disclaimer: this example is modified version of [this](https://docs.rs/csv/*/csv/tutorial/index.html#filter-by-search) one._
+
+Takes a `query` from the command line as input and outputs only the
```
Comment (budziq) [inline src/encoding.md]: > Takes a `query` from the command line as input and outputs only the
well its not true anymore ;)

File: src/encoding.md [hunk 3]
```diff
@@ -543,6 +544,60 @@ Ashley-Zurich-92";
 # quick_main!(run);
 ```
 
+[ex-csv-filter]: #ex-csv-filter
+<a name="ex-csv-filter"></a>
+## Filter CSV records matching a predicate
+
+[![csv-badge]][csv] [![cat-encoding-badge]][cat-encoding]
+
+_Disclaimer: this example is modified version of [this](https://docs.rs/csv/*/csv/tutorial/index.html#filter-by-search) one._
```
Comment (budziq) [inline src/encoding.md]: I would move the disclaimer to bottom of the text or example and reword the text to:
"_Disclaimer: this example has been adapted from the [csv crate tutorial](https://docs.rs/csv/*/csv/tutorial/index.html#filter-by-search) ._"

File: src/encoding.md (hunk 2 above)
Comment (sotocodes) [inline src/encoding.md]: ops 🤦‍

**Comment by sotocodes** [comment]
Hey @budziq, I made the changes but it seems that AppVeyor does not like them :)
> ping me next time :)

roger that!

**Comment by budziq** [comment]
nicely done!

**Comment by sotocodes** [comment]
thanks for the help :)

**Comment by budziq** [comment]
no problem. keep 'em coming :)

**Pull request merged**